mijia-homie/src/main.rs

```diff
@@ -396,42 +417,64 @@ async fn connect_sensor_with_id(
     Ok(())
 }
 
-async fn connect_and_subscribe_sensor_or_disconnect<'a>(
+// TODO: Does this already exist somewhere?
+/// Return the first Ok element of the stream, or a vector of all errors if they all fail.
+async fn first_ok<V, E>(stream: impl Stream<Item = Result<V, E>>) -> Result<V, Vec<E>> {
+    let mut stream = Box::pin(stream);
+    let mut errors = vec![];
+    while let Some(res) = stream.next().await {
+        match res {
+            Ok(v) => return Ok(v),
+            Err(e) => errors.push(e),
+        }
+    }
+    Err(errors)
+}
+
+async fn connect_and_subscribe_sensor_or_disconnect(
     session: &MijiaSession,
-    id: &DeviceId,
-) -> Result<(), eyre::Report> {
-    session
-        .bt_session
-        .connect(id)
-        .await
-        .wrap_err_with(|| format!("connecting to {:?}", id))?;
+    name: &str,
+    ids: Vec<DeviceId>,
+) -> Result<DeviceId, eyre::Report> {
+    // Try to connect to the ids in turn, and get the first one that succeeds. If they all fail then
+    // return an error.
+    let id = first_ok(
+        stream::iter(ids)
+            .then(move |id| async move { session.bt_session.connect(&id).await.map(|()| id) }),
+    )
```

Comment:
just use `for id in ids {... session.bt_session.connect(&id).await ...}` here. No need to mess around with stream combinators anymore, now that we can .await things in a loop.

Reply:
Done.